### ISSUE TITLE
Help wanted: owner needed for windows installer

### ISSUE THREAD

**Issue by jaap-karssenberg** [zim-desktop-wiki/zim-desktop-wiki#639]
A new owner is needed for the windows installer.

See https://lists.launchpad.net/zim-wiki/msg04781.html

**Comment by SuperCuber** [comment]
Hey, I'm currently in the process of trying to build v0.70rc1 on windows, encountered the dependency hell. I will update if I manage to do it. Wish me luck.

**Comment by jaap-karssenberg** [comment]
Good luck !

**Comment by SuperCuber** [comment]
I have figured out how to build and run it. This is my notes from the process. TLDR at the bottom that includes steps that (hopefully) should work. Turns out the dependency hell was much shallower than expected if I use the installer.
[ZimGtk3.txt](https://github.com/zim-desktop-wiki/zim-desktop-wiki/files/2856293/ZimGtk3.txt)

TLDR
====
1. Install python 3.3 portable (not 3.7, does not work with that) from [here](https://sourceforge.net/projects/winpython/files/WinPython_3.3/3.3.5.9/WinPython-64bit-3.3.5.9.exe/download).
2. Use [this blog](https://blogs.gnome.org/kittykat/2014/01/29/developing-gtk-3-apps-with-python-on-windows/) which leads to [this installer](https://sourceforge.net/projects/pygobjectwin32/files/pygi-aio-3.24.1_rev1-setup_049a323fe25432b10f7e9f543b74598d4be74a39.exe/download).
3. Run the installer, point it at `WinPython-64bit-3.3.5.9/python-3.3.5.amd64`.
4. Select the following in each of the 3 package lists:
	a. GTK+
	b. Nothing
	c. GIR (Not sure if neccesary)
5. Everything should complete successfully.
6. Run the portable python shell and use `import gi` to test. If it works, everything went well.
7. In my case, I used the following commands inside a git bash shell to get it to recognize the python installation:
```
ln -s /d/Downloads/WinPython-64bit-3.3.5.9/python-3.3.5.amd64/python.exe /usr/bin/python3.exe
ln -s /d/Downloads/WinPython-64bit-3.3.5.9/python-3.3.5.amd64/python.exe /usr/bin/python.exe
```
8. Restart the shell, and use ./setup.py build. Should complete successfully.

**Comment by jaap-karssenberg** [comment]
In my testing I could run from source and build just following the
instructions from https://www.gtk.org/download/windows.php to install the
msys environment. Only tricky part is that you need to open the right msys
shell to see it work (e.g. the 64 bit shell for 64 bit packages).

Hoever being able to run "./setup.py build" does not yet mean we can build
an installer like the one Brendan build for the python2/Gtk2 branch. For
that we also will need some installer framework to package the build. But
it is a first necessary step.

-- Jaap

**Comment by SuperCuber** [comment]
OK. Any pointers on how to do that? For example, what does the python2 version use?

**Comment by SuperCuber** [comment]
I messed around with it, and right now the problem is that `python.exe  setup.py py2exe --dist-dir 'windows\\build\\ZimDesktopWiki'` results in `ImportError: atk`, and a similar one for `pangocairo` if atk is removed. Both of them are required [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/master/setup.py#L288). Am I missing some dependency? Those packages don't exist in pip

**Comment by jaap-karssenberg** [comment]
@SuperCube: Brendan documented the process for python2/gtk2 in the
"windows" folder in the source. Unfortunately this is not straightforward
compatible with python3/Gtk3. This is the part where we need help to sort
it out and create a new build process. E.g. the packages you refer to are
related to Gtk2 and thus not relevant and are part of what needs to be
updated.

**Comment by jaap-karssenberg** [comment]
This page contains some links to strategies to turn an msys environment
into an installer: https://github.com/msys2/msys2/wiki/Distributing

**Comment by lukas-jaros** [comment]
First of all, thanks for great piece of sofware! ZIM seems to be really usefull application for managing and maintaining operational documentation that changes every while.

I was following SuperCuber's guide to build zim for windows. I was able to successfuly run ZIM by using WinPython 3.3 portable + PYGI-AIO 3.24.1 after running "python setup.py build" and running python zim.py.

While trying to prepare win executable I run into several issues with py2exe, so I tried pyinstaller instead. I managed to go through initial issues but then run into ZIM crashing/exiting without error. Everything was run on WIN 10 machine.

Steps I did:
1. Install WinPython 3.3 portable
2. Install PYGI-AIO 3.24.1 with following packages:
**Gnome/Freedesktop**
base
atk
GDK-Pixbuf
GDL
GEGLGTK -- probably not needed
GSpell -- not needed
GTK+
GTKSourceVIew -- not needed
Pango
**Non-gnome**
GTKSpell -- not needed
**Dev**
GIR
Additional Tools -- not needed
Most of "not needed" packages I installed since I wanted to be able to run ZIM with some plugins pre-installed.

3. pip install PyInstaller==3.3 (higher versions seem to not like Python 3.3
4. copy zim.py into zim_start.py (without this it seems pyinstaller choose to not use zim.py at all)
5. pyinstaller zim_start.py
6. Modify zim_start.spec file with the following:
add: options = [ ('v', None, 'OPTION') ]
change: exe = EXE(pyz,
          a.scripts,
		  options,
          exclude_binaries=True,
          name='zim_start',
          debug=True,
          strip=False,
          upx=False,
          console=True )
7. Afterwards I run "pyinstaller zim_start.spec" and dist\zim_start was created
8. Afterwards I copied gdkpixbuf and gmodule into gi_typelibs (.typelib)
9. At the end I run dist\zim_start\zim_start.exe
10. Seems to load without errors, but it immediately closes. Seems it silently fails while/after "import zim.fs" or import "zim.errors" (file zim\applications.py) and I could not find any reason for that so far. sxstrace shows nothing, nor does the logfile (%TEMP%\zim.exe.txt)

I will continue with investigation in next few days and will let you know. Any hints are highly appreciated :)

**Comment by SuperCuber** [comment]
@lukas-jaros I have only tried py3.3 and py3.7 and 3.7 didn't work. I don't remember really what broke sadly and I don't have the time to investigate, but perhaps versions inbetween could work. I also ran into the pip issue but I actually did upgrade it and broke it 😆 took a little effort to fix

**Comment by jaap-karssenberg** [comment]
This page has some links on building a full installer for windows based on the msys environment:  https://github.com/msys2/msys2/wiki/Distributing

Styrene seems to work as advertised
- will need patch to also install from source into package tree
- code is easy to follow, can insert hook after installing pacman packages
- simple command to call setup.py and let it install in the same root

**Comment by jaap-karssenberg** [comment]
Email from previous maintainer with some work notes here: https://lists.launchpad.net/zim-wiki/msg04848.html

**Comment by qrsp** [comment]
Thanks for everyone's advice. Follow the previous maintainer's note, I can run zim without error and follow jaap-karssenberg's recommendation, building a portable version with Styrene.
I made a [cfg file](https://github.com/zim-desktop-wiki/zim-desktop-wiki/files/3995675/zim.zip) by [gtk3-examples.cfg](https://github.com/achadwick/styrene/blob/master/gtk3-examples.cfg) and [mypaint.cfg](https://github.com/mypaint/mypaint/blob/master/windows/styrene/mypaint.cfg). It's just a prototype and I have not fully tested yet.
Here are some tips for who want to change the theme and sourceview style.
Change theme:
1. download theme from https://www.gnome-look.org/
2. extract to /mingw64/share/themes
3. create settings.ini in /mingw64/etc/gtk-3.0
```ini
[Settings]
gtk-theme-name=THEMENAME
```
Change sourceview style:
1. download style from https://wiki.gnome.org/Projects/GtkSourceView/StyleSchemes
2. extract to /mingw64/share/gtksourceview-3.0/styles
3. edit [sourceview.py](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/881ce9a459f9575ab53f5e8510e1f7d8e64c10ac/zim/plugins/sourceview.py#L150)
```python
class SourceViewBuffer(_bufferclass):

	def __init__(self, attrib, text):
		GtkSource.Buffer.__init__(self)
		self.set_highlight_matching_brackets(True)
		# scheme
		ss = GtkSource.StyleSchemeManager()
		sd = ss.get_scheme('STYLENAME')
		GtkSource.Buffer.set_style_scheme(self, sd)


		# end
		if attrib['lang']:
			self._set_language(attrib['lang'])
```
I hope this is helpful.

**Comment by CrushCrush1954** [comment]
Thanks to the comments on this thread, to the notes of Brendan Kidwell, the old maintainer and of course to the huge efforts of Jaap Karssenberg, I propose you a summary to build zimwiki, unluckily not final.
 
The installation does not come to an end, **something with pyinstaller is missing**. I install as well several variants of the same packages.

## Install packages
### MYSY2  [(Reference)](https://www.glump.net/howto/desktop/install-zim-0-70-or-later-in-windows)
[Download msys2, currently msys2-x86_64-20190524.exe](https://www.msys2.org)
Open MSYS2 from Start Menu

#### Update packages
`pacman -Suy` (if Pacman package is updated, close terminal, reopen MSYS2 and rerun the command to finish the update process)

#### Install Packages
[Reference for using packages with msys2](https://www.msys2.org/wiki/Using-packages/)

### Install Pip [(Reference)](https://packages.msys2.org/package/mingw-w64-x86_64-python-pip)
`pacman -S mingw-w64-x86_64-python-pip`

### Install Python3, GTK3, Python bindings
`pacman -S mingw-w64-x86_64-gtk3 mingw-w64-x86_64-python3 mingw-w64-x86_64-python3-gobject`

### Configure System environment variables [(Reference)](https://www.glump.net/howto/desktop/install-zim-0-70-or-later-in-windows)
    C:\msys64\usr\bin
    C:\msys64\mingw32\bin
    C:\msys64\mingw64\bin

### Python Windows package
[Download from official website](http://www.python.org/download/)

**Note**: Py2exe  [(Reference)](https://pypi.org/project/py2exe/) is incompatible with python3.4 and later ([source](https://stackoverflow.com/questions/41578808/python-indexerror-tuple-index-out-of-range-when-using-py2exe/42310168)), pyinstaller recommended

Install pyinstaller with pip: `pip install pyinstaller`

### Install GTK3 Runtime for Windows
Download current version, currently gtk3-runtime-3.24.14-2020-02-21-ts-win64.exe from [GTK for Windows Runtime](https://github.com/tschoonj/GTK-for-Windows-Runtime-Environment-Installer/releases)

### Visual C++ 2008 Redistributable Package (x86)
Download exactly [this version](http://www.microsoft.com/en-us/download/details.aspx?id=29)

### Git repository
In MSYS2, install git package with `pacman -S git` 

execute `git clone --depth=1 https://github.com/zim-desktop-wiki/zim-desktop-wiki.git` 

**Alternative**: download [zim wiki current revision of master branch](https://github.com/zim-desktop-wiki/zim-desktop-wiki/archive/master.zip) and unzip it to C:\msys64\home\Administrator\zim-desktop-wiki   (remove -master from the directory name or adjust following lines)

### jpeg62.dll
- Download jpeg-6b-4-bin from [Website](http://gnuwin32.sourceforge.net/packages/jpeg.htm)
- Extract jpeg-6b-4-bin\bin\jpeg62.dll and copy it in C:\msys64\home\Administrator\zim-desktop-wiki\windows\lib  (create lib directory)
**Note:** assume current user is local built-in administrator

## changes to C:\msys64\home\Administrator\zim-desktop-wiki\windows\build_win32.py
- l.31, replace
GTK_ROOT = path.join(get_python_lib(), "gtk-2.0", "runtime")
with GTK_ROOT = "C:\Program Files\GTK3-Runtime Win64"

- l.70, Comment out  check py2exe
- l.120 Replace gtk2 with gtk3:
f = open(path.join(EXE_ROOT, r"etc\gtk-3.0\gtkrc"), "w")

# Build
Start program mingw64 (msys2 program group) and execute following commands:
cd zim-desktop-wiki
python windows/build_win32.py
python windows/make_installer_win32.py
python windows/make_installer_win32_portable.py

**Note:** Delete build directory (C:\msys64\home\Administrator\zim-desktop-wiki\windows\build) in case of error before restarting the command.

As mentioned above, the setups will be created, but the main program, zim.exe is not build correctly.

**Comment by fmos** [comment]
I have a working installer and pushed the script for generating it to my branch. @jaap-karssenberg do you want to host the setup.exe somewhere or should I rather host it on my website?

**Comment by jaap-karssenberg** [comment]
Cool!

Can you send a pull request to have the script in the main branch? Ideally
I would want to run it every time as part of my release process.

Until then I can either host your installer or refer to your website -
whatever you prefer.

Regards,

Jaap

**Comment by fmos** [comment]
Yes, I plan to send you a pull request. Before that I want to Add missing dependencies of packaged plugins (e.g. gtkspell, dot, etc).Integrate a Gtk3 theme that provides a more native look and feel to Windows users.


I don't think you can run that script as part of your release process, as PyInstaller does not cross-compile as far as I know. I assume your release process does not involve running a Windows instance?!

I think it makes more sense to host the installer on your official website in the long run. 

You can download the current version here (link works for one week): https://fmos.at/jirafeau/f.php?h=28YKyT16&d=1

For now this is purely for testing purposes and considered beta.

Cheers,
Fabian

Am 08.05.2020 07:34:36, schrieb Jaap Karssenberg:

> 

Cool!
> 
> 

Can you send a pull request to have the script in the main branch? Ideally
> 

I would want to run it every time as part of my release process.
> 
> 

Until then I can either host your installer or refer to your website -
> 

whatever you prefer.
> 
> 

Regards,
> 
> 

Jaap
> 
> 
> 

On Thu, May 7, 2020 at 11:03 PM Fabian Stanke <notifications@github.com>
> 

wrote:

**Comment by fmos** [comment]
Btw: If you want, you can assign me this issue.

**Comment by seihtam** [comment]
I have been using the following Gtk3 theme the last 6 month to make Zim look native on Windows 10, and it have been working great: https://github.com/B00merang-Project/Windows-10

Here are some notes i wrote to myself while figuring out how to install the theme:
* If installed with MSYS2 then gtk config is located at: msys64\mingw64\etc\gtk-3.0\settings.ini
* Themes are located in msys64\mingw64\share\themes
* Disable GTK window decorations to get windows native: setx -m GTK_CSD 0 (https://developer.gnome.org/gtk3/stable/gtk-running.html)
* Apparently gtk3 can't get a native windows theme right now, but gtk2 have MS-Windows http://gtk.10911.n7.nabble.com/gtk-on-windows-10-td93484.html
* But this theme gets close to native https://github.com/B00merang-Project/Windows-10 (to install place in theme folder and change gtk-theme-name in settings.ini)

**Comment by jaap-karssenberg** [comment]
Will give the installer a try on monday.

If the installer compiles in a virtual environment I could look into using
that for the release process, or I could setup a windows box specific for
this.

-- Jaap

**Comment by chrysos349** [comment]
@fmos i tried your `build.sh`. i got `No such file or directory: 'src/zim.spec'`. what did i do wrong? see the full log attached.
[log1.txt](https://github.com/zim-desktop-wiki/zim-desktop-wiki/files/4613234/log1.txt)

**Comment by fmos** [comment]
> @fmos i tried your `build.sh`. i got `No such file or directory: 'src/zim.spec'`. what did i do wrong? 
Thank you @chrysos349 for testing! It seems I forgot to commit the spec file. Sorry, my fault. I will fix this tonight.

**Comment by fmos** [comment]
Hi @chrysos349 ! I have updated my branch. If you want to test, the updated commit is 472c969550fcb4c8fe914547ed9ace4a1a0e94c0

**Comment by chrysos349** [comment]
@fmos Thanks for the update. I built the installer, and installed the app successfully.  The main toolbar has icons missing. I tried `--skip-theme`. The same problem. See the screenshot.
![zim-missing-icons](https://user-images.githubusercontent.com/63523825/81764028-45516e00-94d9-11ea-9df2-d710e881236c.png)

**Comment by fmos** [comment]
This should be fixed with the present version. I also reopened the pull request.

**Comment by jaap-karssenberg** [comment]
Thanks!  I'm planning a new release in 2-3 weeks (most features are in
develop branch, want to use a bit longer to test and merge number of
smaller pulls) would be a good timing to announce this new installer with
that release.

-- Jaap

**Comment by jaap-karssenberg** [comment]
One more question - I tried installing today on my work laptop but without
admin rights I cannot run the installer - is there a way to make the
installer run "portable" ?

Thanks,

Jaap

**Comment by fmos** [comment]
Yes, I think it should be straightforward to build a portable using PyInstaller. As far as I understand, this unpacks into %TEMP% and runs from there each time you start it. So it would probably be more efficient for you to install somewhere you can write to.

@jaap-karssenberg  is there a way to find out which Gtk icons must be shipped with Zim? At the moment the package includes the entire Adwaita icon set (Gtk 3 default). This needs roughly one third of the total installed disk space, which seems disproportionate to me.

**Comment by jaap-karssenberg** [comment]
Hi Fabian,

To get the icons used, a grep like this should work - needs bit of post
processing to translate to direct icon names.

    grep -REoh "icon=['\"]?gtk-\w+['\"]?'|Gtk.STOCK_\w*"
zim-desktop-wiki/zim/ | sort --unique

For the portable installer. I think you can install in a fixed location
without admin rights as long as you don't touch the registry. Typically an
installer would just prompt for the location and I choose a location where
I do have rights.

Regards,

Jaap

**Comment by fmos** [comment]
Hi Jaap!
Thanks für the grep-magic! This does not seems to reveal the icons used for the toolbar (strong, italics, etc). I think those are selected by the `ui_format_actions` in lines 153 ff of pageview.py Strangely, though, the stock id there (`gtk-bold`) does not correspond to an icon image. What Comes closest is the "format-text-bold-symbolic", which is an Adwaita-icon. However this still Looks different than the button in Zim. I'm lost. Any idea where the Image data for those toolbar buttons comes from?

**Comment by jaap-karssenberg** [comment]
Yes you are right, those actions are missed by the grep but are defined in
pageview.py from line 153.

I'm not 100% sure, but I think Gtk translates the (deprecated) stock IDs to
names according this standard list:
https://specifications.freedesktop.org/icon-naming-spec/icon-naming-spec-latest.html

This means you probably are looking for "text-format-bold" (without the
"-symbolic" postfix).

There is an application "gtk3-icon-browser" which defaults installs with
Gtk that is very helpful in searching icon names.

Hope this helps. If not, I'll need to clean up the code and get rid of the
stock IDs altogether and replace with the new icon names.

-- Jaap

**Comment by fmos** [comment]
Thanks! For now I used the warnings about missing Icons logged to the console. The version in my windows-deploy branch packages only the needed Icons and locales. It also supports portable install. It needs some more testing though. Feedback is very welcome.

**Comment by jaap-karssenberg** [comment]
@fabian just released version 0.73, I'll give the installer another test
run and try compiling it myself on this version in the coming week or two

Regards,

Jaap

**Comment by fmos** [comment]
@jaap-karssenberg Thanks for the new release!
Concerning the Windows deployment, I was wondering whether Zim should be started with --standalone by default. I think the debug log to the temporary file only works with this option. Are there any downsides that you would expect?

**Comment by jaap-karssenberg** [comment]
Yes downside is that you can open the same notebook multiple times causing
conflicts. Also opening a new notebook will be slower because the whole
process is initialized again.

Is there something I can fix to make the logging work?

Op zo 7 jun. 2020 15:33 schreef Fabian Stanke <notifications@github.com>:

**Comment by fmos** [comment]
I'll see if I can find out how to fix the logging. Meanwhile I've rebased my branch to 0.73.

**Comment by fmos** [comment]
OK, it seems to work just fine. I didn't get that the log file is switched during startup.

**Comment by CrushCrush1954** [comment]
@fmos,
Thanks to your hard work, I have successfully build Zim-Wiki on a clean Windows 10 (1909 dev VM, as provided by Microsoft) (*),

-> On Start, Zim wiki proposes to create a notebook within "~/Notebooks\Notes". This path does not mean much for Windows users as it links to 'C:\Users\User\Notebooks\Notes' and not to 'Documents' which is linked per default to C:\Users\User\Documents (Sub-directory) on disk.
=> Can you change the default path to 'Documents' instead of ~ ? You may use the environment variable '%USERPROFILE%\Documents'


*: What I did in case someone would like to double-check:
- Download msys2 64-bit from https://www.msys2.org
- Follow instructions on https://msys2.org (Updates package, restart msys2 x64 and re-update)
- Run pacman -S git to install git
- Run git clone https://github.com/fmos/zim-desktop-wiki.git
- git checkout windows-deploy
- Change directory with cd zim-desktop-wiki/windows
- Execute ./build.sh: all required dependencies are downloaded and installer is build
- Installer zim-desktop-wiki-0.73.0-setup.exe to be found within C:\msys64\home\User\zim-desktop-wiki\windows\dist

**Comment by jaap-karssenberg** [comment]
Now I also managed to build my own version in a virtualbox environment.
Will proceed to merge with 0.73.1 branch and publish on the zim main site.

Thanks Fabian - looks great !

-- Jaap

**Comment by jaap-karssenberg** [comment]
Installer is published on https://zim-wiki.org/downloads/

Any issues with the installer can be reported as new issues. Please put clearly "windows installer" in the subject.

Regards,

Jaap
